iroha/src/sumeragi.rs

```diff
@@ -836,6 +888,32 @@ pub mod message {
     }
 
     impl Message {
+        /// # Errors
+        /// Asserts specific instruction number of instruction constraint
+        fn check_instruction_len(&self, max_instruction_number: usize) -> Result<()> {
+            use Message::*;
+
+            match self {
+                BlockCreated(block) => block.check_instruction_len(max_instruction_number),
+                BlockSigned(block) => block.check_instruction_len(max_instruction_number),
+                BlockCommitted(block) => block.check_instruction_len(max_instruction_number),
+                TransactionForwarded(tx_forward) => {
+                    tx_forward.check_instruction_len(max_instruction_number)
+                }
+                NoTransactionReceiptReceived(no_tx_receipt) => {
+                    no_tx_receipt.check_instruction_len(max_instruction_number)
+                }
+
+                CommitTimeout(_) | TransactionReceived(_) | BlockCreationTimeout(_) => Ok(()),
+            }
+        }
+
+        /// # Errors
+        /// Asserts specific instruction number of instruction constraint
+        pub fn validate(&self, max_instruction_number: usize) -> Result<()> {
+            self.check_instruction_len(max_instruction_number)
+        }
+
```

Comment:
There is no need to check instruction length for all messages. Instruction length and transaction byte size should only be checked by Validators at `BlockCreated` message, so that they sign - vote only for the blocks which have proper transactions.

Reply:
In general the check should happen at this point in code:
```
Role::ValidatingPeer => {
                    if !self.block.is_empty()
                        && !self
                            .block
                            .has_committed_transactions(&*sumeragi.world_state_view.read().await)
                        && sumeragi.latest_block_hash == self.block.header().previous_block_hash
                        && sumeragi.number_of_view_changes
                            == self.block.header().number_of_view_changes
                        && sumeragi.block_height + 1 == self.block.header().height
                    {
```

This is where validator decides if it will sign the block. If the block fails before this point some of the important events will not be send.

This part can be definitely improved and probably moved out to a separate function, but conceptually it should happen here.